xarray/core/indexing.py

```diff
@@ -531,6 +514,55 @@ def __repr__(self):
                 (type(self).__name__, self.array, self.key))
 
 
+class LazilyVectorizedIndexedArray(ExplicitlyIndexedNDArrayMixin):
+    """Wrap an array to make vectorized indexing lazy.
+    """
+
+    def __init__(self, array, key):
+        """
+        Parameters
+        ----------
+        array : array_like
+            Array like object to index.
+        key : VectorizedIndexer
+        """
+        if isinstance(key, (BasicIndexer, OuterIndexer)):
+            self.key = VectorizedIndexer(
+                _outer_to_vectorized_indexer(key.tuple, array.shape))
+        else:
+            self.key = _arrayize_vectorized_indexer(key, array.shape)
+        self.array = as_indexable(array)
+        self.order = np.arange(self.ndim)
+
+    @property
+    def shape(self):
+        return np.broadcast(*self.key.tuple).shape
+
+    def __array__(self, dtype=None):
+        return np.asarray(self.array[self.key], dtype=None)
```

Comment:
Shouldn't this be `np.asanyarray()`? Could this interfere with dask arrays? Pint support?

Reply:
This should be okay, neither xarray or dask support using masked arrays under the hood. See  #1194.